File: Android/src/build-backups/kapt3-20260412-000124/stubs/debug/selfgemma/talk/feature/roleplay/roles/RoleCatalogViewModel.java
package selfgemma.talk.feature.roleplay.roles;

@kotlin.Metadata(mv = {2, 2, 0}, k = 1, xi = 48, d1 = {"\u0000n\n\u0002\u0018\u0002\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0002\b\u0003\n\u0002\u0018\u0002\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0002\b\u0003\n\u0002\u0010 \n\u0002\u0018\u0002\n\u0000\n\u0002\u0010\u000e\n\u0002\b\u0005\n\u0002\u0010\u0002\n\u0002\b\u0004\n\u0002\u0010\b\n\u0000\n\u0002\u0010\u0011\n\u0002\u0010\u0000\n\u0002\b\u0002\b\u0007\u0018\u00002\u00020\u0001B;\b\u0007\u0012\b\b\u0001\u0010\u0002\u001a\u00020\u0003\u0012\u0006\u0010\u0004\u001a\u00020\u0005\u0012\u0006\u0010\u0006\u001a\u00020\u0007\u0012\u0006\u0010\b\u001a\u00020\t\u0012\u0006\u0010\n\u001a\u00020\u000b\u0012\u0006\u0010\f\u001a\u00020\r\u00a2\u0006\u0004\b\u000e\u0010\u000fJ\f\u0010\u0017\u001a\b\u0012\u0004\u0012\u00020\u00190\u0018J*\u0010\u001a\u001a\u00020\u001b2\u0006\u0010\u001c\u001a\u00020\u001b2\u0006\u0010\u001d\u001a\u00020\u001b2\n\b\u0002\u0010\u001e\u001a\u0004\u0018\u00010\u001bH\u0086@\u00a2\u0006\u0002\u0010\u001fJ\u000e\u0010 \u001a\u00020!2\u0006\u0010\u001c\u001a\u00020\u001bJ\u000e\u0010\"\u001a\u00020!2\u0006\u0010#\u001a\u00020\u001bJ+\u0010$\u001a\u00020\u001b2\b\b\u0001\u0010%\u001a\u00020&2\u0012\u0010\'\u001a\n\u0012\u0006\b\u0001\u0012\u00020)0(\"\u00020)H\u0002\u00a2\u0006\u0002\u0010*R\u000e\u0010\u0002\u001a\u00020\u0003X\u0082\u0004\u00a2\u0006\u0002\n\u0000R\u000e\u0010\u0004\u001a\u00020\u0005X\u0082\u0004\u00a2\u0006\u0002\n\u0000R\u000e\u0010\u0006\u001a\u00020\u0007X\u0082\u0004\u00a2\u0006\u0002\n\u0000R\u000e\u0010\b\u001a\u00020\tX\u0082\u0004\u00a2\u0006\u0002\n\u0000R\u000e\u0010\n\u001a\u00020\u000bX\u0082\u0004\u00a2\u0006\u0002\n\u0000R\u000e\u0010\f\u001a\u00020\rX\u0082\u0004\u00a2\u0006\u0002\n\u0000R\u0014\u0010\u0010\u001a\b\u0012\u0004\u0012\u00020\u00120\u0011X\u0082\u0004\u00a2\u0006\u0002\n\u0000R\u0017\u0010\u0013\u001a\b\u0012\u0004\u0012\u00020\u00120\u0014\u00a2\u0006\b\n\u0000\u001a\u0004\b\u0015\u0010\u0016\u00a8\u0006+"}, d2 = {"Lselfgemma/talk/feature/roleplay/roles/RoleCatalogViewModel;", "Landroidx/lifecycle/ViewModel;", "appContext", "Landroid/content/Context;", "dataStoreRepository", "Lselfgemma/talk/data/DataStoreRepository;", "roleRepository", "Lselfgemma/talk/domain/roleplay/repository/RoleRepository;", "createRoleplaySessionUseCase", "Lselfgemma/talk/domain/roleplay/usecase/CreateRoleplaySessionUseCase;", "importStRoleCardFromUriUseCase", "Lselfgemma/talk/domain/roleplay/usecase/ImportStRoleCardFromUriUseCase;", "compileRuntimeRoleProfileUseCase", "Lselfgemma/talk/domain/roleplay/usecase/CompileRuntimeRoleProfileUseCase;", "<init>", "(Landroid/content/Context;Lselfgemma/talk/data/DataStoreRepository;Lselfgemma/talk/domain/roleplay/repository/RoleRepository;Lselfgemma/talk/domain/roleplay/usecase/CreateRoleplaySessionUseCase;Lselfgemma/talk/domain/roleplay/usecase/ImportStRoleCardFromUriUseCase;Lselfgemma/talk/domain/roleplay/usecase/CompileRuntimeRoleProfileUseCase;)V", "feedbackState", "Lkotlinx/coroutines/flow/MutableStateFlow;", "Lselfgemma/talk/feature/roleplay/roles/RoleCatalogUiState;", "uiState", "Lkotlinx/coroutines/flow/StateFlow;", "getUiState", "()Lkotlinx/coroutines/flow/StateFlow;", "getSessionPersonaOptions", "", "Lselfgemma/talk/feature/roleplay/roles/SessionPersonaOptionUiState;", "createSession", "", "roleId", "modelId", "personaSlotId", "(Ljava/lang/String;Ljava/lang/String;Ljava/lang/String;Lkotlin/coroutines/Continuation;)Ljava/lang/Object;", "deleteRole", "", "importStRoleCard", "uri", "appString", "resId", "", "args", "", "", "(I[Ljava/lang/Object;)Ljava/lang/String;", "app_debug"})
@dagger.hilt.android.lifecycle.HiltViewModel()
public final class RoleCatalogViewModel extends androidx.lifecycle.ViewModel {
    @org.jetbrains.annotations.NotNull()
    private final android.content.Context appContext = null;
    @org.jetbrains.annotations.NotNull()
    private final selfgemma.talk.data.DataStoreRepository dataStoreRepository = null;
    @org.jetbrains.annotations.NotNull()
    private final selfgemma.talk.domain.roleplay.repository.RoleRepository roleRepository = null;
    @org.jetbrains.annotations.NotNull()
    private final selfgemma.talk.domain.roleplay.usecase.CreateRoleplaySessionUseCase createRoleplaySessionUseCase = null;
    @org.jetbrains.annotations.NotNull()
    private final selfgemma.talk.domain.roleplay.usecase.ImportStRoleCardFromUriUseCase importStRoleCardFromUriUseCase = null;
    @org.jetbrains.annotations.NotNull()
    private final selfgemma.talk.domain.roleplay.usecase.CompileRuntimeRoleProfileUseCase compileRuntimeRoleProfileUseCase = null;
    @org.jetbrains.annotations.NotNull()
    private final kotlinx.coroutines.flow.MutableStateFlow<selfgemma.talk.feature.roleplay.roles.RoleCatalogUiState> feedbackState = null;
    @org.jetbrains.annotations.NotNull()
    private final kotlinx.coroutines.flow.StateFlow<selfgemma.talk.feature.roleplay.roles.RoleCatalogUiState> uiState = null;
    
    @javax.inject.Inject()
    public RoleCatalogViewModel(@dagger.hilt.android.qualifiers.ApplicationContext()
    @org.jetbrains.annotations.NotNull()
    android.content.Context appContext, @org.jetbrains.annotations.NotNull()
    selfgemma.talk.data.DataStoreRepository dataStoreRepository, @org.jetbrains.annotations.NotNull()
    selfgemma.talk.domain.roleplay.repository.RoleRepository roleRepository, @org.jetbrains.annotations.NotNull()
    selfgemma.talk.domain.roleplay.usecase.CreateRoleplaySessionUseCase createRoleplaySessionUseCase, @org.jetbrains.annotations.NotNull()
    selfgemma.talk.domain.roleplay.usecase.ImportStRoleCardFromUriUseCase importStRoleCardFromUriUseCase, @org.jetbrains.annotations.NotNull()
    selfgemma.talk.domain.roleplay.usecase.CompileRuntimeRoleProfileUseCase compileRuntimeRoleProfileUseCase) {
        super();
    }
    
    @org.jetbrains.annotations.NotNull()
    public final kotlinx.coroutines.flow.StateFlow<selfgemma.talk.feature.roleplay.roles.RoleCatalogUiState> getUiState() {
        return null;
    }
    
    @org.jetbrains.annotations.NotNull()
    public final java.util.List<selfgemma.talk.feature.roleplay.roles.SessionPersonaOptionUiState> getSessionPersonaOptions() {
        return null;
    }
    
    @org.jetbrains.annotations.Nullable()
    public final java.lang.Object createSession(@org.jetbrains.annotations.NotNull()
    java.lang.String roleId, @org.jetbrains.annotations.NotNull()
    java.lang.String modelId, @org.jetbrains.annotations.Nullable()
    java.lang.String personaSlotId, @org.jetbrains.annotations.NotNull()
    kotlin.coroutines.Continuation<? super java.lang.String> $completion) {
        return null;
    }
    
    public final void deleteRole(@org.jetbrains.annotations.NotNull()
    java.lang.String roleId) {
    }
    
    public final void importStRoleCard(@org.jetbrains.annotations.NotNull()
    java.lang.String uri) {
    }
    
    private final java.lang.String appString(@androidx.annotation.StringRes()
    int resId, java.lang.Object... args) {
        return null;
    }
}

File: Android/src/app/src/main/java/selfgemma/talk/feature/roleplay/roles/RoleCatalogViewModel.kt
package selfgemma.talk.feature.roleplay.roles

import android.content.Context
import androidx.annotation.StringRes
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import dagger.hilt.android.lifecycle.HiltViewModel
import dagger.hilt.android.qualifiers.ApplicationContext
import javax.inject.Inject
import android.util.Log
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.launch
import selfgemma.talk.data.DataStoreRepository
import selfgemma.talk.R
import selfgemma.talk.domain.roleplay.model.RoleCard
import selfgemma.talk.domain.roleplay.model.availablePersonaSlotIds
import selfgemma.talk.domain.roleplay.model.selectPersonaSlot
import selfgemma.talk.domain.roleplay.model.snapshotSelectedPersona
import selfgemma.talk.domain.roleplay.repository.RoleRepository
import selfgemma.talk.domain.roleplay.usecase.CompileRuntimeRoleProfileUseCase
import selfgemma.talk.domain.roleplay.usecase.CreateRoleplaySessionUseCase
import selfgemma.talk.domain.roleplay.usecase.ImportStRoleCardFromUriUseCase

private const val TAG = "RoleCatalogViewModel"

data class RoleCatalogUiState(
  val loading: Boolean = true,
  val builtInRoles: List<RoleCard> = emptyList(),
  val customRoles: List<RoleCard> = emptyList(),
  val statusMessage: String? = null,
  val errorMessage: String? = null,
)

data class SessionPersonaOptionUiState(
  val slotId: String,
  val name: String,
  val descriptionPreview: String,
  val avatarUri: String? = null,
  val isDefault: Boolean = false,
  val isCurrent: Boolean = false,
)

@HiltViewModel
class RoleCatalogViewModel
@Inject
constructor(
  @ApplicationContext private val appContext: Context,
  private val dataStoreRepository: DataStoreRepository,
  private val roleRepository: RoleRepository,
  private val createRoleplaySessionUseCase: CreateRoleplaySessionUseCase,
  private val importStRoleCardFromUriUseCase: ImportStRoleCardFromUriUseCase,
  private val compileRuntimeRoleProfileUseCase: CompileRuntimeRoleProfileUseCase,
) : ViewModel() {
  private val feedbackState = MutableStateFlow(RoleCatalogUiState(loading = false))

  val uiState: StateFlow<RoleCatalogUiState> =
    combine(roleRepository.observeRoles(), feedbackState) { roles, feedback ->
      RoleCatalogUiState(
        loading = false,
        builtInRoles = roles.filter { it.builtIn },
        customRoles = roles.filterNot { it.builtIn },
        statusMessage = feedback.statusMessage,
        errorMessage = feedback.errorMessage,
      )
    }
      .stateIn(
        scope = viewModelScope,
        started = SharingStarted.WhileSubscribed(5_000),
        initialValue = RoleCatalogUiState(),
      )

  fun getSessionPersonaOptions(): List<SessionPersonaOptionUiState> {
    val profile = dataStoreRepository.getStUserProfile().ensureDefaults()
    val availableSlotIds = profile.availablePersonaSlotIds()
    val defaultSlotId = profile.defaultPersonaId?.trim().takeUnless { it.isNullOrBlank() }
    val currentSlotId = profile.resolvedUserAvatarId()
    return availableSlotIds
      .map { slotId ->
        val slotProfile = profile.selectPersonaSlot(slotId)
        SessionPersonaOptionUiState(
          slotId = slotId,
          name = slotProfile.userName,
          descriptionPreview = slotProfile.personaDescription.firstNonBlankLine(),
          avatarUri = slotProfile.activeAvatarUri,
          isDefault = slotId == defaultSlotId,
          isCurrent = slotId == currentSlotId,
        )
      }.sortedWith(
        compareByDescending<SessionPersonaOptionUiState> { it.isDefault }
          .thenByDescending { it.isCurrent }
          .thenBy { it.name.lowercase() }
          .thenBy { it.slotId },
      )
  }

  suspend fun createSession(roleId: String, modelId: String, personaSlotId: String? = null): String {
    val selectedPersonaProfile =
      dataStoreRepository.getStUserProfile().snapshotSelectedPersona(personaSlotId)
    Log.d(
      TAG,
      "create session roleId=$roleId modelId=$modelId personaSlotId=${selectedPersonaProfile.userAvatarId} personaName=${selectedPersonaProfile.userName}",
    )
    return createRoleplaySessionUseCase(
      roleId = roleId,
      modelId = modelId,
      userProfile = selectedPersonaProfile,
    ).id
  }

  fun deleteRole(roleId: String) {
    viewModelScope.launch {
      roleRepository.deleteRole(roleId)
    }
  }

  fun importStRoleCard(uri: String) {
    viewModelScope.launch {
      runCatching {
        importStRoleCardFromUriUseCase.importFromUri(uri = uri)
      }
        .onSuccess { importedRole ->
          roleRepository.saveRole(compileRuntimeRoleProfileUseCase(importedRole))
          feedbackState.update {
            it.copy(
              statusMessage = appString(R.string.role_catalog_status_st_imported),
              errorMessage = null,
            )
          }
        }
        .onFailure { error ->
          feedbackState.update {
            it.copy(
              statusMessage = null,
              errorMessage = error.message ?: appString(R.string.role_catalog_error_st_import_failed),
            )
          }
        }
    }
  }

  private fun appString(@StringRes resId: Int, vararg args: Any): String {
    return appContext.getString(resId, *args)
  }
}

private fun String.firstNonBlankLine(): String {
  return lineSequence().map(String::trim).firstOrNull { it.isNotBlank() }.orEmpty()
}
